### PR TITLE
DFR-3425 Refactor notification logic second solution

### PR DESCRIPTION
**Pull request by hamzahtahirhmcts**
### Jira link

See [DFR-3425](https://tools.hmcts.net/jira/browse/DFR-3425)

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/Reviewable.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.draftorders.review.OrderStatus;
+
+import java.time.LocalDateTime;
+
+public interface Reviewable {
+    LocalDateTime getSubmittedDate();
+
+    OrderStatus getOrderStatus();
+}
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/Reviewable.java:11]: end line

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrderAboutToSubmitHandlerTest.java [hunk 2]
```diff
@@ -282,4 +284,33 @@ private static Stream<Arguments> provideAgreedDraftOrders() {
             Arguments.of(uadoWithNull, List.of(), List.of())
         );
     }
+
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrderAboutToSubmitHandlerTest.java:287]: remove an extra blank line.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java [hunk 3]
```diff
@@ -66,11 +72,19 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(
             finremCallbackRequest.getEventType(), finremCallbackRequest.getCaseDetails().getId());
 
         FinremCaseDetails caseDetails = finremCallbackRequest.getCaseDetails();
-        NotificationRequest judgeNotificationRequest = draftOrdersNotificationRequestMapper.buildJudgeNotificationRequest(caseDetails);
-        notificationService.sendContestedReadyToReviewOrderToJudge(judgeNotificationRequest);
+        String caseReference = String.valueOf(caseDetails.getId());
+        boolean isAgreedLatest = isLatestUploadAnAgreedDraftOrder(caseDetails.getData(), caseReference);
+
+        if (isAgreedLatest) {
```
[the remark this answers is not in the document]
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java:78]: I reckon so.. We only want to send a notification for agreed orders/PSAs.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/Reviewable.java (hunk 1 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/Reviewable.java]: [dc5f6a7](https://github.com/hmcts/finrem-case-orchestration-service/pull/1978/commits/dc5f6a71b44a89b6224515c9033f0995a309b847)

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrderAboutToSubmitHandlerTest.java (hunk 2 above)
Comment (hamzahtahirhmcts) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrderAboutToSubmitHandlerTest.java:287]: [dc5f6a7](https://github.com/hmcts/finrem-case-orchestration-service/pull/1978/commits/dc5f6a71b44a89b6224515c9033f0995a309b847)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java [hunk 4]
```diff
@@ -66,11 +72,19 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(
...
+            DraftOrdersReview review = getDraftOrderReviewWithLatestSubmissionDate(caseDetails.getData()
+                .getDraftOrdersWrapper());
+            LocalDate hearingDate = review.getHearingDate();
```
[the remark this answers is not in the document]
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java:82]: [2a399d5](https://github.com/hmcts/finrem-case-orchestration-service/pull/1978/commits/2a399d542c934b36113869b2e3520141cf99c74c)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/draftorders/upload/UploadDraftOrdersSubmittedHandler.java:78]: [ca0b318](https://github.com/hmcts/finrem-case-orchestration-service/pull/1978/commits/ca0b31898fbd5a99a0bbb8509abc089abb20e038) 

I've updated the tests to check for notifications on both paths now.

**Comment by so99y** [comment]
Clone the test case from James' branch - https://github.com/hmcts/finrem-case-orchestration-service/pull/1949/files#diff-7d22c35d9011c771032594cb1b1a34f474cd3763ba0d99e1ea48605ce706e7cb

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/mapper/DraftOrdersNotificationRequestMapper.java [hunk 5]
```diff
@@ -17,22 +14,17 @@
...
+    public NotificationRequest buildJudgeNotificationRequest(FinremCaseDetails caseDetails, LocalDate hearingDate, String judge) {
 
         NotificationRequest judgeNotificationRequest = new NotificationRequest();
         judgeNotificationRequest.setCaseReferenceNumber(String.valueOf(caseDetails.getId()));
-        judgeNotificationRequest.setHearingDate(dateFormatter.format(date));
-        judgeNotificationRequest.setNotificationEmail(agreedDraftOrder.getJudge());
-        judgeNotificationRequest.setApplicantName(Objects.toString(caseData.getFullApplicantName()));
+        judgeNotificationRequest.setHearingDate(dateFormatter.format(hearingDate));
+        judgeNotificationRequest.setNotificationEmail(judge);
+        judgeNotificationRequest.setApplicantName(Objects.toString(caseDetails.getData().getFullApplicantName()));
```
[the remark this answers is not in the document]
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/mapper/DraftOrdersNotificationRequestMapper.java]: [4863faa](https://github.com/hmcts/finrem-case-orchestration-service/pull/1978/commits/4863faa6bbe14860d6efde722406f552e66e9048)

**Review by so99y — APPROVED** [review]

**Review by jthmcts — APPROVED** [review]

**Pull request merged**